### PR TITLE
HIVE-27493: HMS optimizes the SQL query index of getPartitionsByFilter method

### PR DESCRIPTION
**Pull request by cxzl25**
### What changes were proposed in this pull request?
Add `PART_KEY_VAL` index to table `PARTITION_KEY_VALS`

### Why are the changes needed?
Improve performance

#### Current
```
[pool-4-thread-1] DEBUG metastore.MetaStoreDirectSql: Direct SQL query in 352.763678ms + 0.026622ms, the query is [select "PARTITIONS"."PART_ID" from "PARTITIONS"  inner join "TBLS" on "PARTITIONS"."TBL_ID" = "TBLS"."TBL_ID"     and "TBLS"."TBL_NAME" = ?   inner join "DBS" on "TBLS"."DB_ID" = "DBS"."DB_ID"      and "DBS"."NAME" = ? inner join "PARTITION_KEY_VALS" "FILTER0" on "FILTER0"."PART_ID" = "PARTITIONS"."PART_ID" and "FILTER0"."INTEGER_IDX" = 0 inner join "PARTITION_KEY_VALS" "FILTER1" on "FILTER1"."PART_ID" = "PARTITIONS"."PART_ID" and "FILTER1"."INTEGER_IDX" = 1 where ( (("FILTER0"."PART_KEY_VAL" = ?) and ("FILTER1"."PART_KEY_VAL" = ?)) )] 
```
#### Add index
```
[pool-4-thread-2] DEBUG metastore.MetaStoreDirectSql: Direct SQL query in 13.770586ms + 0.016063ms
```

### Does this PR introduce _any_ user-facing change?
No

### Is the change a dependency upgrade?
No

### How was this patch tested?
Validation in production environment

### CODE REVIEW

File: standalone-metastore/metastore-server/src/main/sql/derby/hive-schema-4.0.0-beta-1.derby.sql [hunk 1]
```diff
@@ -306,6 +306,8 @@ CREATE INDEX "APP"."CONSTRAINTS_PARENT_TBL_ID_INDEX" ON "APP"."KEY_CONSTRAINTS"(
 
 CREATE INDEX "APP"."CONSTRAINTS_CONSTRAINT_TYPE_INDEX" ON "APP"."KEY_CONSTRAINTS"("CONSTRAINT_TYPE");
 
+CREATE INDEX "APP"."PARTITION_KEY_VALS_IDX" ON "APP"."PARTITION_KEY_VALS"("PART_KEY_VAL");
```
Comment (zhangbutao) [inline standalone-metastore/metastore-server/src/main/sql/derby/hive-schema-4.0.0-beta-1.derby.sql:309]: If we change the latest version schema, we also should make sure the privious schema to upgrade smoothly.
That is to say, we need to add this change in `upgrade-4.0.0-alpha-2-to-4.0.0-beta-1.derby.sql`.

Maybe you can refer to HIVE-19605 to add the change.

**Review by zhangbutao — COMMENTED** [review]
Thanks @cxzl25 for this change. This looks a good improvment.
BTW, could you elaborate how the PR improves query? e.g. What's the type of quey?

**Comment by cxzl25** [comment]
> could you elaborate how the PR improves query? e.g. What's the type of quey

In the case where we use filter conditions to prune partitions, such as `get_partitions_by_filter` RPC.
It is mainly filtered with the field `PARTITION_KEY_VALS` of `PARTITION_KEY_VALS`, but it does not have an index, which leads to poor performance, especially if a table has more partitions.

```sql
SELECT PARTITIONS."PART_ID"
FROM "PARTITIONS"
    INNER JOIN TBLS
    ON PARTITIONS."TBL_ID" = TBLS."TBL_ID"
        AND TBLS."TBL_NAME" = ?
    INNER JOIN DBS
    ON TBLS."DB_ID" = DBS."DB_ID"
        AND DBS."NAME" = ?
    INNER JOIN PARTITION_KEY_VALS "FILTER0"
    ON FILTER0."PART_ID" = PARTITIONS."PART_ID"
        AND FILTER0."INTEGER_IDX" = 0
    INNER JOIN PARTITION_KEY_VALS "FILTER1"
    ON FILTER1."PART_ID" = PARTITIONS."PART_ID"
        AND FILTER1."INTEGER_IDX" = 1
WHERE FILTER0."PART_KEY_VAL" = ?
    AND FILTER1."PART_KEY_VAL" = ?
```

**Review by zhangbutao — APPROVED** [review]
+1 non-binding lgtm.

**Comment by zhangbutao** [comment]
Ping @ayushtkn @dengzhhu653 @deniskuzZ who may be interested in this small but good change. Thanks.

**Comment by deniskuzZ** [comment]
@cxzl25 thanks for the patch!
Could you please update the sql script versions, cause beta1 was already released?

Patch LGTM, however, 1 question: as we already have PRIMARY KEY (`PART_ID`,`INTEGER_IDX`) why do we need an extra index on `PART_ID`?

**Comment by cxzl25** [comment]
> update the sql script versions

Done

> Patch LGTM, however, 1 question: as we already have PRIMARY KEY (`PART_ID`,`INTEGER_IDX`) why do we need an extra index on `PART_ID`?

The index I added is `PART_KEY_VAL`, not `PART_ID`.

**Comment by deniskuzZ** [comment]
> > update the sql script versions
> 
> Done
> 
> > Patch LGTM, however, 1 question: as we already have PRIMARY KEY (`PART_ID`,`INTEGER_IDX`) why do we need an extra index on `PART_ID`?
> 
> The index I added is `PART_KEY_VAL`, not `PART_ID`.

that's true, i was just wondering why we need `PART_ID` if it's already part of PK

**Comment by cxzl25** [comment]
> i was just wondering why we need `PART_ID` if it's already part of PK.

I checked that the index for `PART_ID` exists from very early versions, it is really not necessary.

```sql
  PRIMARY KEY (`PART_ID`,`INTEGER_IDX`),
  KEY `PARTITION_KEY_VALS_N49` (`PART_ID`),
```

> do we know what is the impact on write operations

If the discussion is about the impact of newly added indexes on writing, it can almost be ignored.
This index has been added in production for several years and it works fine.

**Pull request closed without merging**